stdlib/public/core/Substring.swift.gyb

```diff
@@ -19,7 +19,17 @@ extension String {
   ///
   /// - Complexity: O(*n*), where *n* is the length of `substring`.
   public init(_ substring: Substring) {
-    self = String(substring._slice)
+    let x = substring._wholeString
+    let start = substring.startIndex
+    let end = substring.endIndex
+    let u16 = x._core[start.encodedOffset..<end.encodedOffset]
+    if start.samePosition(in: x.unicodeScalars) != nil
+    && end.samePosition(in: x.unicodeScalars) != nil {
+      self = String(_StringCore(u16))
```

Comment:
Yes, all that crud should be got rid of.  The thing to remember is that ICU is going to do validation and correction for us whether we want it or not, and no user is going to be upset by seeing invalid code sequences, so there's no sense in paying twice.